design-proposals/delete-changes-on-nncp-delete.md

```diff
@@ -0,0 +1,188 @@
+# Delete nmstate configuration from the system on NNCP deletion
+
+## Summary
+
+We want kubernetes-nmstate to delete changes made by an NNCP when that NNCP is deleted.
+
+## TLDR
+
+Deleting NNCP will do whatever applying output of `nmstatectl gr` does.
+
+## Motivation
+
+Users find it confusing that deleting the NNCP does not remove the changes it applied. The current workflow for removing such a configuration is to apply the same NNCP but with `state: absent`. As we have `nmstatectl gr` that generates a newtwork state reverting the desired state, we are able to make the behavior of NNCP easier for users.
+
+### User Stories
+
+1) As a cluster admin I apply NNCP that creates a VLAN. I want that deleting this NNCP removes the VLAN.
+
+2) As a cluster admin I apply NNCP that defines static IP address on an interface. I want that deleting this NNCP removes this IP address from the interface.
+
+### Goals
+
+1) We want users to have a consistent behavior between `nmstatectl gr` and NNCP deletion. The behavior when deleting NNCP should be aligned with what would happen if user generated manifest using `nmstatectl gr` and applied it as NNCP.
+
+2) We want kubernetes-nmstate to use its health probes to block removal of NNCP if that would break connectivity to the node.
+
+### Non-Goals
+
+1) We are not supporting removal/revert of the configuration when NNCP is modified.
+
+2) We are not taking snapshot of network configuration before NNCP was applied in order to revert to that snapshot. We are only deleting changes that specific NNCP applies.
```

Comment:
Say an interface is DHCP-configured, and an NNCP changes it to a static IP. What happens when the NNCP is deleted? Does it revert to DHCP or to no IP configuration at all?

Or an SR-IOV PF with 4 VFs. NNCP increases to 6 VFs and then NNCP is deleted. Outcome is 4 VFs or no VFs?